### PR TITLE
report validation

### PR DESCRIPTION
**Pull request by Bjvanminnen**
We have a lot of different places that call sendReport in lots of different ways. It can be very hard to follow these, and understand what is actually reported. This leads to some weird things like a field always being either the string `"true"` or the value `false`. There's another field where sometimes it is an object, and sometimes it is a json string of the object.

This is an attempt to (a) document some of the weirdness (b) assert on the current state (such that if someone adds a new report, it needs to follow the same schema).

Another thing that I noticed was that we send the server a bunch of fields in our report that it doesn't need/use. This PR makes it so that we only send the server what it will use. This change is a little bit riskier, in that it's possible I got things wrong and have stopped sending the server some field it does actually care about.

There are also a couple of small changes I made in what we actually provide to sendReport. My initial goal was to make sendReport more consistent. I ultimately decided this was going to be riskier, and instead focused on documenting the ways it wasn't consistent.

### CODE REVIEW

File: apps/src/calc/calc.js [hunk 1]
```diff
@@ -677,7 +677,6 @@ Calc.execute = function () {
   var reportData = {
     app: 'calc',
     level: level.id,
-    builder: level.builder,
```
Comment (Bjvanminnen) [inline apps/src/calc/calc.js:680]: wasnt being used anywhere

File: apps/src/code-studio/initApp/loadApp.js [hunk 2]
```diff
@@ -117,7 +117,6 @@ export function setupApp(appOptions) {
         report.callback = appOptions.report.callback;
         trackEvent('Activity', 'Lines of Code', window.script_path, report.lines);
       }
-      report.scriptName = appOptions.scriptName;
```
Comment (Bjvanminnen) [inline apps/src/code-studio/initApp/loadApp.js:120]: No longer used

File: apps/src/code-studio/reporting.js [hunk 3]
```diff
@@ -39,13 +40,173 @@ reporting.getLastServerResponse = function () {
...
+        }
+        break;
+      case 'containedLevelResultsInfo':
+        validateType('containedLevelResultsInfo', value, 'object');
+        break;
+      default:
+        // Eventually we'd probably prefer to throw here, but I don't have enough
+        // confidence that this validation is 100% correct to start breaking things
+        // if it isnt.
+        console.error(`Unexpected report key '${key}' of type '${typeof report[key]}'`);
```
Comment (Bjvanminnen) [inline apps/src/code-studio/reporting.js:177]: Worth noting that I had an earlier version of this branch where we did throw here, and UI tests all passed.

**Comment by Bjvanminnen** [comment]
I've gone through most of the levels in allthethings and submitted, giving me some confidence that this is now an accurate representation, however it's very possible there are still edge cases that will end up resulting in console.errors now.

File: apps/src/code-studio/reporting.js [hunk 4]
```diff
@@ -39,13 +40,173 @@ reporting.getLastServerResponse = function () {
...
+ * Validate that the provided field on our report object is one of the given
+ * type
+ * @param {string} key
+ * @param {*} value
+ * @param {string} type
+ */
+function validateType(key, value, type) {
+  let typeIsValid = false;
+  if (type === 'array') {
+    typeIsValid = typeIsValid || _.isArray(value);
```
Comment (islemaster) [inline apps/src/code-studio/reporting.js]: Why not [`Array.isArray`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/isArray)?

Comment (Bjvanminnen) [reply apps/src/code-studio/reporting.js]: Only because I wasn't familiar with such a thing :) Will change.

File: apps/src/code-studio/reporting.js [hunk 5]
```diff
@@ -39,13 +40,173 @@ reporting.getLastServerResponse = function () {
...
+  validateReport(report);
+
   // jQuery can do this implicitly, but when url-encoding it, jQuery calls a method that
   // shows the result dialog immediately
   var queryItems = [];
   for (var key in report) {
-    if (report.hasOwnProperty(key) && key !== 'onComplete') {
-      queryItems.push(key + '=' + report[key]);
+    if (report.hasOwnProperty(key)) {
+      if (serverFields.includes(key)) {
```
Comment (islemaster) [inline apps/src/code-studio/reporting.js]: This seems like an appropriate place to use [`_.pick`](https://lodash.com/docs/4.17.4#pick)

**Comment by Bjvanminnen** [comment]
We do not afaik. Adding that would possibly be a reasonable thing to do, but I think I've perhaps already spent too much time on this, so I think I'm going to refrain from adding that for now.
